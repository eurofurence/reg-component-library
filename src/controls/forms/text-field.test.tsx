/** @jsxImportSource @emotion/react */

import TextField from './text-field'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ChangeEventHandler } from 'react'
import * as fc from 'fast-check'

describe('TextField', () => {
	it('renders correctly', () => {
		fc.assert(fc.property(fc.fullUnicodeString(), fc.fullUnicodeString(), (label, placeholder) => {
			const tree = renderer
				.create(<TextField name="username" label={label} placeholder={placeholder}/>)
				.toJSON()

			expect(tree).toMatchSnapshot()
		}), { numRuns: 20, seed: 2562346 })
	})

	it('has a label', () => {
		const { queryByLabelText } = render(
			<TextField name="username" label="Username" placeholder=""/>,
		)

		expect(queryByLabelText('Username')).toBeTruthy()
	})

	it('enters focus when clicked directly', () => {
		const { getByRole } = render(
			<TextField name="username" label="Username" placeholder=""/>,
		)

		const textField = getByRole('textbox')

		expect(textField).not.toHaveFocus()

		userEvent.click(textField)

		expect(textField).toHaveFocus()
	})

	it('enters focus when clicked through label', () => {
		const { getByRole, getByText } = render(
			<TextField name="username" label="Username" placeholder=""/>,
		)

		const label = getByText('Username')
		const textField = getByRole('textbox')

		expect(textField).not.toHaveFocus()

		userEvent.click(label)

		expect(textField).toHaveFocus()
	})

	it('can handle unicode input', () => {
		fc.assert(fc.property(fc.fullUnicodeString({ minLength: 1 }), (input) => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLInputElement>>
			const { getByRole } = render(
				<TextField name="username" label="Username" placeholder="" onChange={onChange}/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, input)

			expect(onChange.mock.calls[onChange.mock.calls.length - 1]![0].target.value).toEqual(input)
		}))
	})

	it('can display unicode', () => {
		fc.assert(fc.property(fc.fullUnicodeString({ minLength: 1 }), (input) => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLInputElement>>
			const { getByRole } = render(
				<TextField name="username" label="Username" placeholder="" value={input} onChange={onChange}/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, input)

			expect(textField).toHaveValue(input)
		}))
	})

	describe('when controlled', () => {
		it('reflects the input value at all times (true)', () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLInputElement>>
			const { getByRole } = render(
				<TextField name="username" label="Username" value="test" placeholder="" onChange={onChange}/>,
			)

			const textField = getByRole('textbox')

			expect(textField).toHaveValue('test')

			userEvent.type(textField, 'john-doe')

			expect(textField).toHaveValue('test')
		})

		it('raises onChange events when typed in', () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLInputElement>>
			const { getByRole } = render(
				<TextField name="username" label="Username" value="test" placeholder="" onChange={onChange}/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, 'john-doe')

			expect(onChange).toHaveBeenCalled()
		})

		it('supports readOnly fields', () => {
			const { getByRole } = render(
				<TextField name="username" label="Username" value="test" placeholder="" readOnly/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, 'john-doe')

			expect(textField).toHaveValue('test')
		})
	})

	describe('when uncontrolled', () => {
		it('reflects the typed value', () => {
			const { getByRole } = render(
				<TextField name="username" label="Username" placeholder=""/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, 'john-doe')

			expect(textField).toHaveValue('john-doe')
		})

		it('reflects the defaultValue value originally', () => {
			const { getByRole } = render(
				<TextField name="username" label="Username" placeholder="" defaultValue="uwu"/>,
			)

			expect(getByRole('textbox')).toHaveValue('uwu')
		})

		it('raises onChange events when typed in', () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLInputElement>>
			const { getByRole } = render(
				<TextField name="username" label="Username" placeholder="" onChange={onChange}/>,
			)

			const textField = getByRole('textbox')

			userEvent.type(textField, 'john-doe')

			expect(onChange).toHaveBeenCalled()
		})
	})
})
