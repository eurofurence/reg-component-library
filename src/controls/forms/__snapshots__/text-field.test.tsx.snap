// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField renders correctly 1`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񹇁򆙳"
  />
</label>
`;

exports[`TextField renders correctly 2`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    /򫘍𘅮񆐭
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="󲸽r"
  />
</label>
`;

exports[`TextField renders correctly 3`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder=""
  />
</label>
`;

exports[`TextField renders correctly 4`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񜩢"
  />
</label>
`;

exports[`TextField renders correctly 5`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    􏿼
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񧽋㋔񍂶򑳲"
  />
</label>
`;

exports[`TextField renders correctly 6`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    娱񹮗𛗃𳟁
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񆠟"
  />
</label>
`;

exports[`TextField renders correctly 7`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="valueOf"
  />
</label>
`;

exports[`TextField renders correctly 8`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="l"
  />
</label>
`;

exports[`TextField renders correctly 9`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="򭮀񯧠򷉄򚣙򜜼󷒛񊂳"
  />
</label>
`;

exports[`TextField renders correctly 10`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    "
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="괫󂟥񪽿񷄎񃶟𺸬𩢍"
  />
</label>
`;

exports[`TextField renders correctly 11`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="𮙊񢛬󯨦򷗸󺳍𚡆𤶮򉸋"
  />
</label>
`;

exports[`TextField renders correctly 12`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    __proto__
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="󒸺𩿥񌇉笉󳖷񴥶򡀴󘥄𝹑"
  />
</label>
`;

exports[`TextField renders correctly 13`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    񈯾
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="󐶻Ի󢙟򛪰󇍨"
  />
</label>
`;

exports[`TextField renders correctly 14`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    argumentsr
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="󤈍􏿺"
  />
</label>
`;

exports[`TextField renders correctly 15`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    򭺩󄌄󺨚󁒲󬁀񠩹ㄝ󚛈򃋮󇉠
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="𽏵󿃂􏿺𺃎"
  />
</label>
`;

exports[`TextField renders correctly 16`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    ﳽ򔬞󯦱򿯑𷍘򘃵򷥹򏸍󂒏󗰟
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder=""
  />
</label>
`;

exports[`TextField renders correctly 17`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    $􏿷
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񋽢򸗭񥕆􂶛󌤋"
  />
</label>
`;

exports[`TextField renders correctly 18`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    񣧺򧕾𓞆󷣄𴚏񮋟𙽮򳭆
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="񁸴󌓍'񡩧􉌲򔭋􆓀񺍎񅇡"
  />
</label>
`;

exports[`TextField renders correctly 19`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    󫂄񂺛
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="􏿸"
  />
</label>
`;

exports[`TextField renders correctly 20`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    󕚱󤐀𔖢󾲳򯀯󒟽򾄢􃄂񂑟򡝴
  </div>
  <input
    className="css-16p6ckh"
    name="username"
    placeholder="򊞄􌽑򷔫󄱗𸝩􇡫􎶊𯴏"
  />
</label>
`;
